File: tools/simulator_manager/Sources/LRUSet.swift
/// A `Set` that has a maximum capacity and evicts the least recently used item
// when full.
struct LRUSet<Element: Hashable> {
  private let capacity: Int
  // An array to keep track of the order in which elements were inserted.
  // The first element in the array is the least recently used.
  private var order: [Element] = []

  // A Set to enable fast O(1) membership tests.
  private var storage: Set<Element> = []

  init(capacity: Int) {
    precondition(capacity > 0, "Capacity must be greater than zero.")
    self.capacity = capacity
  }

  // Returns an element that was evicted from the set, or nil if nothing was evicted
  // (including when `element` was already present and just moved to most-recently-used).
  mutating func insert(_ element: Element) -> Element? {
    if storage.contains(element) {
      if let index = order.firstIndex(of: element) {
        order.remove(at: index)
      }

      order.append(element)
      return nil
    }

    let evicted: Element?
    if order.count >= capacity, let oldest = order.first {
      order.removeFirst()
      evicted = storage.remove(oldest)
    } else {
      evicted = nil
    }

    order.append(element)
    storage.insert(element)
    return evicted
  }

  func contains(_ element: Element) -> Bool {
    return storage.contains(element)
  }

  var elements: [Element] {
    return order
  }
}

File: tools/simulator_manager/go/lru_set.go
package main

// LRUSet is a Set that has a maximum capacity and evicts the least recently used item when full.
type LRUSet[T comparable] struct {
	capacity int
	// An array to keep track of the order in which elements were inserted.
	// The first element in the array is the least recently used.
	order []T
	// A map to enable fast O(1) membership tests.
	storage map[T]struct{}
}

func NewLRUSet[T comparable](capacity int) *LRUSet[T] {
	if capacity <= 0 {
		panic("Capacity must be greater than zero.")
	}
	return &LRUSet[T]{
		capacity: capacity,
		order:    make([]T, 0, capacity),
		storage:  make(map[T]struct{}),
	}
}

// Insert returns an element that was evicted from the set.
func (s *LRUSet[T]) Insert(element T) *T {
	var evicted *T
	if _, exists := s.storage[element]; exists {
		// Remove from current position in order
		for i, e := range s.order {
			if e == element {
				old := s.order[i]
				evicted = &old
				s.order = append(s.order[:i], s.order[i+1:]...)
				break
			}
		}
		s.order = append(s.order, element)
	} else {
		if len(s.order) >= s.capacity && len(s.order) > 0 {
			oldest := s.order[0]
			s.order = s.order[1:]
			delete(s.storage, oldest)
			evicted = &oldest
		}
		s.order = append(s.order, element)
		s.storage[element] = struct{}{}
	}
	return evicted
}

func (s *LRUSet[T]) Contains(element T) bool {
	_, exists := s.storage[element]
	return exists
}

func (s *LRUSet[T]) Elements() []T {
	return s.order
}
